pkg/common/kafka/sarama/sarama.go

```diff
@@ -0,0 +1,266 @@
+		if err != nil {
+			return nil, fmt.Errorf("ConfigMap's eventing-kafka value could not be converted to an EventingKafkaConfig struct: %s : %v", err, configMap[constants.EventingKafkaSettingsConfigKey])
+		}
+	}
+
+	if eventingKafkaConfig == nil {
+		// A nil config is not valid; create an empty config so that defaults may be set
+		eventingKafkaConfig = &commonconfig.EventingKafkaConfig{}
+	}
+
+	// Set Some Default Values If Missing
+
+	// Increase The Idle Connection Limits From Transport Defaults If Not Provided (see net/http/DefaultTransport)
+	if eventingKafkaConfig.CloudEvents.MaxIdleConns == 0 {
+		eventingKafkaConfig.CloudEvents.MaxIdleConns = constants.DefaultMaxIdleConns
+	}
+	if eventingKafkaConfig.CloudEvents.MaxIdleConnsPerHost == 0 {
+		eventingKafkaConfig.CloudEvents.MaxIdleConnsPerHost = constants.DefaultMaxIdleConnsPerHost
+	}
+
+	// Default Replicas To 1 If Not Specified
+	if eventingKafkaConfig.Channel.Receiver.Replicas < 1 {
+		eventingKafkaConfig.Channel.Receiver.Replicas = 1
+	}
+	if eventingKafkaConfig.Channel.Dispatcher.Replicas < 1 {
+		eventingKafkaConfig.Channel.Dispatcher.Replicas = 1
+	}
+
+	// Set Default Values For Secret
+	if len(eventingKafkaConfig.Kafka.AuthSecretNamespace) == 0 {
+		eventingKafkaConfig.Kafka.AuthSecretNamespace = system.Namespace()
+	}
+	if len(eventingKafkaConfig.Kafka.AuthSecretName) == 0 {
+		eventingKafkaConfig.Kafka.AuthSecretName = DefaultAuthSecretName
+	}
+
+	return eventingKafkaConfig, nil
+}
+
+// AuthFromSarama creates a KafkaAuthConfig using the SASL settings from
+// a given Sarama config, or nil if there is no SASL user in that config
+func AuthFromSarama(config *sarama.Config) *client.KafkaAuthConfig {
+	// Use the SASL settings from the provided Sarama config only if the user is non-empty
+	if config.Net.SASL.User != "" {
+		return &client.KafkaAuthConfig{
+			SASL: &client.KafkaSaslConfig{
+				User:     config.Net.SASL.User,
+				Password: config.Net.SASL.Password,
+				SaslType: string(config.Net.SASL.Mechanism),
+			},
+		}
+	} else {
+		// If the user is empty, return explicitly nil authentication
+		return nil
+	}
+}
+
+// StringifyHeaders Is A Utility function to convert []byte headers to string ones for logging purposes
+func StringifyHeaders(headers []sarama.RecordHeader) map[string][]string {
+	stringHeaders := make(map[string][]string)
+	for _, header := range headers {
+		key := string(header.Key)
+		stringHeaders[key] = append(stringHeaders[key], string(header.Value))
+	}
+	return stringHeaders
+}
+
+// StringifyHeaderPtrs Is A Pointer-version of the StringifyHeaders function
+func StringifyHeaderPtrs(headers []*sarama.RecordHeader) map[string][]string {
+	stringHeaders := make(map[string][]string)
+	for _, header := range headers {
+		key := string(header.Key)
+		stringHeaders[key] = append(stringHeaders[key], string(header.Value))
+	}
+	return stringHeaders
+}
```

Comment:
can both call into the same "helper" ?

Reply:
I have tried to combine these functions and the result is not any better than these separate ones because of the amount of overhead necessary ("slice of pointers" is different than "slice of RecordHeaders", despite go papering over that and allowing you to write "header.Value" for either).  A Go master might know something better (happy to accept something!) but I threw in the towel.